### PR TITLE
Update styled-components to version 4.4.0

### PR DESCRIPTION
**Pull request by ariya**
This should eliminate many React warnings on deprecated lifecycles methods (e.g. componentWillReceiveProps).

### Pre-requisites

* [x] #17048
* [x] #17402
* [x] #17863
* [x] #17847
* [ ] #17862
* [x] #17890
* [ ] npm local patch

## Tasks

* [x] Upgrade style-components to 4.4.0
* [x] Upgrade grid-styled to 5.5.0
* [x] Patch grid-styled 5
* [x] Fix ClampedText's innerRef
* [ ] Fix Link
* [x] Fix TextInput's innerRef

This is basically a variant of PR #16995, except that the upgraded version is 4.4.0. To check, first remove `--silent` option from Jest in `package.json` and then run the unit tests as usual: `yarn test-unit`.

**Before**

Searching for `Please update the following components:` should give almost 200 occurrences.

**After**

Searching for `Please update the following components:` should give only 19 occurrences.

### CODE REVIEW

**Review by gusaiani — APPROVED** [review]

**Comment by gusaiani** [comment]
Thank you very much, @ariya.

#### More looking forward than about this PR 

We now are getting new warnings of two kinds, related to styled-components.

```js
console.warn node_modules/styled-components/dist/styled-components.cjs.js:1971
      Functions as object-form attrs({}) keys are now deprecated and will be removed in a future version of styled-components. Switch to the new attrs(props => ({})) syntax instead for easier and more powerful composition. The attrs key in question is "mr" on component "Radiostyled__NormalItem".
```

This looks like just moving forward and adapting to the API in v5 of styled-components, no sweat.

And

```js
console.warn node_modules/grid-styled/node_modules/styled-components/dist/styled-components.cjs.js:2221
      It looks like there are several instances of 'styled-components' initialized in this application. This may cause dynamic styles not rendering properly, errors happening during rehydration process and makes you application bigger without a good reason.
```

in my investigation, in v5 we get version interactions between grid-styled and styled-components.

There is an update to grid-styled, it's just renamed @rebass/grid

Now that we use styled-components more often, we could consider writing `Box`, `Flex` and other components currently coming from grid-styled ourselves, and dropping the library altogether.

I love libs, sure, but our uses of Box and Flex could be rethought, extending rather than passing so many props each time we call a component.

To share what I noticed by exploring @rebass-grid, it changes layouts of things. Buttons lose margins and so on.

So it may be a good thing to evaluate the implications of both approaches. 
And also having the visual regression tool available to speed up this evaluation, even if done very manually.
cc @alxnddr

**Comment by gusaiani** [comment]
I could swear tests were green when I approved. My bad.

**Comment by ariya** [comment]
Note that somehow the component's margin/padding has changed. I'm looking into it:

originally in **master**

image


**this PR**

image

File: frontend/test/metabase/scenarios/smoketest/user.cy.spec.js [hunk 1]
```diff
@@ -69,6 +69,8 @@ describe("smoketest > user", () => {
     });
     cy.button("Visualize").click();
 
+    cy.wait(1000); // slide-in animation for the table
```
Comment (ariya) [inline frontend/test/metabase/scenarios/smoketest/user.cy.spec.js:72]: I don't fancy this, though there are already similar waits in this test file. Any better idea @nemanjaglumac ?

Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/smoketest/user.cy.spec.js:72]: Let me take a look. There has to be something else that we can wait for in the UI.
About those other arbitrary waits - we should get rid of them asap.

Comment (nemanjaglumac) [reply frontend/test/metabase/scenarios/smoketest/user.cy.spec.js]: @ariya done :)
https://github.com/metabase/metabase/pull/17056/commits/205c9e8f2441d5c98e61544e73dbee11e6348d77

Comment (ariya) [reply frontend/test/metabase/scenarios/smoketest/user.cy.spec.js]: Thanks @nemanjaglumac :+1:

**Comment by ariya** [comment]
> Note that somehow the component's margin/padding has changed. I'm looking into it:

Look at this particular component (simplified) constructed with `Flex`:

```js
<Flex className="flex-full z1" pr={2} align="center">
  <Box w={1} style={{ maxWidth: 500 }}>
    <SearchBar/>
  </Box>
</Flex>
```

According to [styled-system documentation](https://styled-system.com/getting-started#margin--padding), `pr` is for right padding.

The theme passed to ThemeProvider is as follows:
```
const theme = {
  space: [4, 8, 16, 32, 64, 128],
};
```

With style-components 3.2.6 in the master branch, this results in the following paddings (0px 16px 0px 80px):

image

whereas with style-components 4.4, it becomes slight different paddings (0px 8px 0px 80px):

image

Looks like the value lookup for `pr` (or other padding/margin attributes for that) has changed.

Strangely, if I just double the numbers in `theme` array, there are various other (possibly default?) margins and paddings which mess up the UI even further.

**Next step**: look into the guts of styled-system to see how the margin/padding lookup works.

**Comment by ariya** [comment]
> **Next step**: look into the guts of styled-system to see how the margin/padding lookup works.

The plot thickens. If we only look at styled-components and styled-system, ignoring Box/Flex for now, then it seems that the margin/padding lookups work as expected.

I use these simple test components:
```js
const Container = styled.div`
  ${space}
  background-color: red;
`;

const SearchBox = styled.input`
  ${space}
`
```

and when they are rendered with:
```
        <Container p={1}>
          <SearchBox p={2} m={3} value={"Search"}/>
        </Container>
```

then the paddings and margins for both components are correctly taken from ThemeProvider's theme, both with styled-components v3 and v4.

**Next step**: figure out the role of Box/Flex from grid-styled.

**Comment by ariya** [comment]
> **Next step**: figure out the role of Box/Flex from grid-styled.

Apparently Box/Flex from grid-styled doesn't work correctly, when used in conjunction with style-components 4.4.0 (it works fine obviously with styled-components 3.2.6). Let's say we have the following:

```
<Box m={1}/>
```
then the margin for that component is always 4px, regardless of the theme supplied via ThemeProvider, because of `m={1}`, basically multiplying that number with 4.

After **a lot** of digging and walking through the source code, I found the culprit: an error thrown by styled-component when that Box is rendered:
```
bootstrap:27 Uncaught TypeError: Cannot read property 'rules' of undefined
    at createStyledComponent (styled-components.browser.esm.js:2152)
    at templateFunction (styled-components.browser.esm.js:1352)
    at Object.../../../node_modules/grid-styled/dist/Box.js (Box.js:25)
```

While the above error did not prevent the component to be constructed and rendered correctly, it did however stop the right theming to be applied, particular for the spacing, and therefore causing the Box to adopt its default spacing (the integer multiply of 4px).

Looking at the code of grid-styled even further indicates that, even if the above problem can be rectified, grid-styled 4.1 will still not work with style-components 4.x because it still uses `extend`, which was removed in v4 of styled-components. In fact, we have removed the use of `extend` in our own code to facilitate styled-components upgrade (see PR #17048 for details).

I have tried a few other version bumps of grid-styled, from 4.1.1 all the way to 4.3.3, and they are all still incompatible with style-components 4. Only grid-styled 5 will work without causing any exception to be thrown (the good news), but now the margin/padding is getting worse (the bad news). Icons are somehow missing as well.

image

**Next step**: make grid-styled 5 honors and adopts the theme from ThemeProvider.

**Comment by ariya** [comment]
> **Next step**: make grid-styled 5 honors and adopts the theme from ThemeProvider.

Looks like `Box` from grid-styled 5 works just fine, it adopts whatever theme set through ThemeProvider.
OTOH, `Flex` mysteriously does not work the same way.

Let's take a look at the following JSX, where both `Flex` and `Box` should specify the same margin:
```
      <Flex className="test-flex" m={2}>
        <Box className="test-box" m={2}></Box>
      </Flex>
```

and then we observe the rendered elements via the DevTools. Note that for the `div` created from `Box`, the margin is applied correctly, see the CSS rules in `.figgFS`.

image


However, if we look at the enclosing `div` created from `Flex`, there is no margin rule at all, neither in `.iHZvIS` nor `.hKiLMS`. Very mysterious!

image

**Comment by ariya** [comment]
> OTOH, `Flex` mysteriously does not work the same way.

I've experimented with monkey-patching grid-styled 5 so that the behavior of `Flex` resembles that of `Box`, see commit  5befcd65.

The good news: this gives the proper margin and padding all over the places!
The bad news: some icons in the UI are still missing.

image

**Next step**: solve the case of missing icons.

**Comment by ariya** [comment]
It turned out that the disappearing icons (and also, among others, the incorrect background) was due to the updated Box/Flex doesn't recorgnize "property shorthand" anymore, e.g. `w` (for width), `bg` (for background-color), etc. Once I found that out, the fix for the landing page is easy: simply set the necessary style rules explicitly inside `style` prop. This is illustrated as follows.

Before:
```js
<Flex
  align="center"
  justify="center"
  bg={color("accent4")}
  w="42px"
  style={{ borderRadius: 6, height: 42 }}
  mr={1}
>
  <Icon name="bolt" size={20} className="flex-no-shrink text-white" />
</Flex>
```

After:
```js
<Flex
  align="center"
  justify="center"
  style={{
    backgroundColor: color("accent4"),
    borderRadius: 6,
    width: 42,
    height: 42,
  }}
  mr={1}
>
  <Icon name="bolt" size={20}/>
</Flex>
```

With that, the landing page is now pixel perfect, as expected:

image
Of course, it doesn't solve everything since apparently we have numerous (over a hundred) of Flex usages all over the place. We can obviously systematically change every use of Flex not to use any shorthands like `w` and `bg` anymore, but that'll be a lot of changes and hopefully nothing is missed.

OTOH, personal preference aside, making the CSS rules more explicit like isn't bad at all, since each CSS property set becomes logically grouped and make it easier, e.g. to look for and identify related changes.

*Next step*: Contemplate whether to fix all the callee sites or to find some hacks to keep the shorthand compatibility.

**Comment by ariya** [comment]
@metabase-bot run visual tests

**Comment by ariya** [comment]
@metabase-bot run visual tests

**Comment by ariya** [comment]
@metabase-bot run visual tests

**Comment by ariya** [comment]
@metabase-bot run visual tests

**Comment by ariya** [comment]
This is obsoleted and not needed anymore, since we decided to go with emotion instead: #20265.

**Pull request closed without merging**